### PR TITLE
Fix subquery all subquery context

### PR DESCRIPTION
**Pull request by dgkimura**
If SubqueryAll operator has a Subquery scalar child then we should treat
the subquery context as EsqctxtValue, even if the SubqueryAll comes from
the predicate. Let's use the following example:

```
SELECT * FROM foo WHERE (SELECT a FROM foo limit 1) = ALL(SELECT b FROM bar);

+--CScalarSubqueryAll(=)["b" (8)]
   |--CLogicalGet "bar" ("bar"),
   +--CScalarSubquery["a" (16)]
      +--CLogicalLimit <empty> global
         |--CLogicalGet "foo" ("foo"),
         |--CScalarConst (0)
         +--CScalarCast
            +--CScalarConst (1)
```

Setting subquery context EsqctxtValue signals to transform the scalar
subquery into a left outer apply as opposed to an inner apply in
function FGenerateCorrelatedApplyForScalarSubquery(). A left outer apply
is necessary to preserve non-matching rows in the output (inner apply
would have filtered them out) which are required to correctly determine
the ALL_SUBLINK result.

### CODE REVIEW

File: src/test/regress/sql/qp_correlated_query.sql [hunk 1]
```diff
@@ -192,6 +192,38 @@ select A.i, B.i, C.j from A, B, C where A.j < all ( select C.j from C where not
 explain select A.i, B.i, C.j from A, B, C where A.j = all (select C.j from C where C.j = A.j and not exists (select sum(B.i) from B where C.i = B.i and C.i !=10)) order by A.i, B.i, C.j limit 10;
 select A.i, B.i, C.j from A, B, C where A.j = all (select C.j from C where C.j = A.j and not exists (select sum(B.i) from B where C.i = B.i and C.i !=10)) order by A.i, B.i, C.j limit 10;
 
+-- -- -- --
+-- Test ALL clause with subqueries
+-- -- -- --
+create table qp_csq_all_t1(a int) distributed by (a);
+create table qp_csq_all_t2(b int) distributed by (b);
+
+truncate qp_csq_all_t1, qp_csq_all_t2;
```
Comment (l-wang) [inline src/test/regress/sql/qp_correlated_query.sql:201]: No need to truncate here.

**Review by l-wang — APPROVED** [review]
LGTM with minor nitpicks.

**Comment by l-wang** [comment]
For the sake of my own learning, I played with the test query more and confirmed that Greenplum gets same results as upstream Postgres, so the patch is good. However I don't really understand the semantics of the following queries:

setup:
```sql
create table qp_csq_all_t1(a int) distributed by (a);
create table qp_csq_all_t2(b int) distributed by (b);
insert into qp_csq_all_t1 values (null);
insert into qp_csq_all_t1 values (null);
``` 
query 1: why do we get 2 rows when we have a `limit 1` in the ALL subquery clause?
```
postgres=# select * from qp_csq_all_t1 where (select a from qp_csq_all_t1 limit 1) = all(select b from qp_csq_all_t2);
 a 
---
  
  
(2 rows)
```
query 2: How does the ALL operator work when the left hand side has more than 1 rows?
```
postgres=# select * from qp_csq_all_t1 where (select a from qp_csq_all_t1) = all(select b from qp_csq_all_t2);
 a 
---
  
  
(2 rows)
```

**Comment by l-wang** [comment]
> query 1: why do we get 2 rows when we have a `limit 1` in the ALL subquery clause?
> 
> ```
> postgres=# select * from qp_csq_all_t1 where (select a from qp_csq_all_t1 limit 1) = all(select b from qp_csq_all_t2);
>  a 
> ---
>   
>   
> (2 rows)
> ```

Answering my own question. So really we are evaluating the WHERE clause for each row in t1. Here's a different example:

```sql
create table qp_csq_all_t1(a int) distributed by (a);
create table qp_csq_all_t2(b int) distributed by (b);
insert into qp_csq_all_t1 values (1);
insert into qp_csq_all_t1 values (2);
insert into qp_csq_all_t1 values (1);
```

The same query would return non-deterministic results.
```
postgres=# select * from qp_csq_all_t1 where (select a from qp_csq_all_t1 limit 1) = all(select b from qp_csq_all_t2);
 a 
---
(0 rows)

postgres=# select * from qp_csq_all_t1 where (select a from qp_csq_all_t1 limit 1) = all(select b from qp_csq_all_t2);
 a 
---
 1
 2
(2 rows)
```

I'm still surprised how query 2 in my previous comment works. For this setup, query 2 fails:
```
postgres=# select * from qp_csq_all_t1 where (select a from qp_csq_all_t1) = all(select b from qp_csq_all_t2);
ERROR:  21000: more than one row returned by a subquery used as an expression  (seg1 slice2 127.0.1.1:7003 pid=431367)
LOCATION:  ExecScanSubPlan, nodeSubplan.c:364
```

**Comment by dgkimura** [comment]
On Monday, April 11, 2022 2:41 PM Lei (Alexandra) Wang wrote:

> The same query would return non-deterministic results, depending on what we get from the limit 1 clause.

Yeah, you're right.  I think we need an order by to get deterministic results.

> I'm still surprised how query 2 in my previous comment works. For this setup, query 2 fails:
>
> ```
> postgres=# select * from qp_csq_all_t1 where (select a from qp_csq_all_t1) = all(select b from qp_csq_all_t2);
> ERROR:  21000: more than one row returned by a subquery used as an expression  (seg1 slice2 127.0.1.1:7003 pid=431367)
> LOCATION:  ExecScanSubPlan, nodeSubplan.c:364
> ```

Yeah, I was surprised too.  Your previous comment seems to be a special case for when right subquery is empty then the all expression must be true (even if the left subquery contains more than 1 row).

```
select '{1, 2}' = all('{}');
 ?column? 
----------
 t
(1 row)
```

Interesting though is your second case also works with constants:

```
select '{1, 2}' = all('{1}');
 ?column? 
----------
 f
(1 row)
```

I kinda feel like this behavior is contradicting.  As this is consistent with Postgres, perhaps we can address separate from this PR?

**Comment by l-wang** [comment]
> As this is consistent with Postgres, perhaps we can address separate from this PR?

Agreed. This PR LGTM.

**Pull request merged**